File: Solutions/111_Minimum_Depth_of_Binary/111_ Minimum_Depth_of_Binary.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int MinDepth(TreeNode root) {
        if(root == null)
            return 0;
        
        if(root.left == null && root.right == null)
            return 1;
        if(root.left == null)
            return MinDepth(root.right) + 1;
        if(root.right == null)
            return MinDepth(root.left) + 1;
        
        return Math.Min(MinDepth(root.left), MinDepth(root.right)) + 1;
    }
}

//Most common
public class Solution
{
    public int MinDepth(TreeNode root)
    {
        if (root == null)
        {
            return 0;
        }
        
        if (root.left == null && root.right == null)
        {
            return 1;
        }
        
        int min = int.MaxValue;
        if (root.right != null)
        {
            min = Math.Min(MinDepth(root.right), min);
        }
        
        if (root.left != null)
        {
            min = Math.Min(MinDepth(root.left), min);
        }
        
        return min + 1;
    }
}

//Third fastest
public class Solution {
    
    private int m_MinimumDepth = int.MaxValue;
    
    public int MinDepth(TreeNode root) {
        FindNearestEndNode(root, 1);
        if (m_MinimumDepth == int.MaxValue)
            return 0;
            
        return m_MinimumDepth;
    }
    
    private void FindNearestEndNode(TreeNode node, int level)
    {
        if (node == null)
            return;
                  
        var result = level;
        
        if (node.left != null)
        {
            FindNearestEndNode(node.left, level + 1);
        }
        if (node.right != null)
        {
            FindNearestEndNode(node.right, level + 1);
        }        
        
        if (node.left == null && node.right == null && level < m_MinimumDepth)
            m_MinimumDepth = level;
    }
}

//Second fastest
public class Solution {
    public int MinDepth(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null) return 1 + MinDepth(root.right);
        if (root.right == null) return 1 + MinDepth(root.left);
        return 1 + Math.Min(MinDepth(root.left), MinDepth(root.right));
    }

//Fastest
public class Solution {
    public int MinDepth(TreeNode root) {
    int leftDepth = 0;
    int rightDepth = 0;
    if (root == null) return 0;
        
    leftDepth = MinDepth(root.left);
    rightDepth = MinDepth(root.right);
     
    if (leftDepth!=0 && rightDepth!=0)
        return Math.Min(leftDepth,rightDepth) + 1;    
    else if (leftDepth ==0)
        return rightDepth + 1;    
    else
        return leftDepth + 1;    
    }
}

File: Solutions/111_Minimum_Depth_of_Binary/111_ Minimum_Depth_of_Binary.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

//Fourth fastest
var minDepth = function(root) {    
    if(root === null) {
        return 0;
    }
    if (root.left === null && root.right === null) {
      return 1;
    }
    let minimumDepth = Number.MAX_VALUE;
    if(root.left !== null) {
        let leftDepth = minDepth(root.left);
        minimumDepth = Math.min(leftDepth, minimumDepth);
    }
    if(root.right !== null) {
        let rightDepth = minDepth(root.right);
        minimumDepth = Math.min(rightDepth, minimumDepth);
    }
    return minimumDepth+1;
   
};

//Third fastest
var minDepth = function(root) {
    if (!root) return 0
    if (!root.left && !root.right) return 1

    let depth = Number.MAX_SAFE_INTEGER

    if (root.left) {
        depth = Math.min(minDepth(root.left), depth)
    }

    if (root.right) {
        depth = Math.min(minDepth(root.right), depth)
    }

    return depth + 1
};

//Second fastest
var minDepth = function(root) {
    if (!root) return 0;
     let queue = [] 
    queue.push(root)
    let minDepth = 0;
    while(queue.length > 0){
        minDepth++ 
        let level = queue.length
      for(let i = 0; i < level; i++){
          let node = queue.shift()
          
          if(!node.left  && !node.right ){
              return minDepth
          }
          
           if(node.right !== null){
              queue.push(node.right)
          }
          if(node.left !== null){
              queue.push(node.left)
          }
          
      }
    }
    return minDepth
};

//Fastest
var minDepth = function(root) {
    if (!root)  return 0;
    
    let depth = 1
    let queue = [root]
    while (queue.length > 0) {
        let len = queue.length
        for (let i = 0; i < len; i++) {
            let node = queue.shift()
            if (node === null) return 0
            if (node.left === null && node.right === null) return depth
            if (node.left !== null) queue.push(node.left)
            if (node.right !== null) queue.push(node.right)
        }
        depth++
    }
    return depth
};